packages/next/client/use-intersection.tsx

```diff
@@ -34,43 +34,46 @@ export function useIntersection<T extends Element>({
 
   const unobserve = useRef<Function>()
   const [visible, setVisible] = useState(false)
-  const [root, setRoot] = useState(rootRef ? rootRef.current : null)
-  const setRef = useCallback(
-    (el: T | null) => {
+  const elementRef = useRef<T | null>(null)
+
+  useEffect(() => {
+    if (hasIntersectionObserver) {
       if (unobserve.current) {
         unobserve.current()
         unobserve.current = undefined
       }
 
       if (isDisabled || visible) return
 
+      const el = elementRef.current
       if (el && el.tagName) {
         unobserve.current = observe(
           el,
           (isVisible) => isVisible && setVisible(isVisible),
-          { root, rootMargin }
+          { root: rootRef?.current, rootMargin }
         )
       }
-    },
-    [isDisabled, root, rootMargin, visible]
-  )
-
-  const resetVisible = useCallback(() => {
-    setVisible(false)
-  }, [])
 
-  useEffect(() => {
-    if (!hasIntersectionObserver) {
+      return () => {
+        unobserve.current?.()
+        unobserve.current = undefined
+      }
+    } else {
       if (!visible) {
         const idleCallback = requestIdleCallback(() => setVisible(true))
         return () => cancelIdleCallback(idleCallback)
       }
     }
-  }, [visible])
+  }, [isDisabled, rootMargin, rootRef, visible])
+
+  const resetVisible = useCallback(() => {
+    setVisible(false)
+  }, [])
+
+  const setRef = useCallback((el: T | null) => {
+    elementRef.current = el
```

Comment:
Do we still need to store the element reference in a `useState()` since when `setRef` is called we update the reference here but we aren't re-calling the `useEffect` above so we won't observe the new ref unless the parent component re-renders?

Or are we changing this hook to have the assumption that the ref passed to `setRef` should always only change when the parent re-renders?

Reply:
> Do we still need to store the element reference in a `useState()` since when `setRef` is called we update the reference here but we aren't re-calling the `useEffect` above so we won't observe the new ref unless the parent component re-renders?
>
> Or are we changing this hook to have the assumption that the ref passed to setRef should always only change when the parent re-renders?

Here is what I thought. `setRef` here is only supposed to be called after component mounts. In `<Image />` and `<Link />` case, it is called inside the [DOM's callback ref](https://reactjs.org/docs/refs-and-the-dom.html#callback-refs).

And the assumption here is that the DOM node (stored in the ref) itself would remain the same before the component unmounts.

If we use state here, when React mounts the `<Image />` (or `<Link />`), the state needs to be updated after the component mounts, which would cause an extra update, when we only need to re-execute the effect.

We could, however, add `elementRef.current` to the `useEffect`'s dependency array. This might be an anti-pattern, but it does make sure when the `elementRef` is updated, the effect will be re-executed.

Reply:
After reading the React documentation again, now I thought updating a state inside a callback ref is not a bad idea after all: [How can I measure a DOM node?
 - Hooks FAQ](https://reactjs.org/docs/hooks-faq.html#how-can-i-measure-a-dom-node).

I will change it to store the element reference in a state instead.

Reply:
How about a test to count renders? https://github.com/vercel/next.js/pull/37447#pullrequestreview-1004444781

Reply:
> How about a test to count renders? [#37447 (review)](https://github.com/vercel/next.js/pull/37447#pullrequestreview-1004444781)

@styfle I still have no idea how to test this though. It seems that there is no realistic way to test against component update counts.

One way is to patch the React reconciler through `SECRET_INTERNAL_DO_NOT_USE_OR_YOU_WILL_BE_FIRED`.

Another way is to add a global variable, which will have to live inside the `next/image` component, introducing completely useless bytes in the production.

Reply:
How about `globalThis.componentCount` so that its global and doesnt live inside a component and therefor outside of React's jurisdiction

Reply:
> How about `globalThis._testRenderCount` so that its global and doesnt live inside a component and therefor outside of React's jurisdiction

`globalThis._testRenderCount` does live outside of React, but **the modification** toward the `globalThis._testRenderCount` **has to** live inside the `next/image` component (and probably has to be inside a `useEffect`), which is what I am really concerned about.